### PR TITLE
MOS: Emit more efficient DEC/CMP pattern, DCP on 6502X, for globals.

### PR DESCRIPTION
**Pull request by asiekierka**
This implements the optimization discussed in https://github.com/llvm-mos/llvm-mos/issues/360 . This is also the first (work in progress) optimization to use an ~~illegal~~ unofficial opcode.

The benefit:

* With unofficial opcodes: `8 * (n + 1)` bits: `{4, 5} * n - 2` bytes saved, `{7, 8} * n - 2` cycles saved (savings for >= 16 bits)
* Without unofficial opcodes: `8 * (n + 1)` bits: `2 * n - 2` bytes saved, `2 * n - 2` cycles saved (savings for >= 24 bits, equal for 16 bits)

TODO:

- [x] It's disappointing that it only applies to globals currently. The same optimization could apply for values which are stored in imaginary registers, as they are also memory locations. I'm not yet sure how to achieve this, though.
- [x] Get some sleep and re-review the code before sending for third-party review.

### CODE REVIEW

**Comment by asiekierka** [comment]
There we go. I'm mostly satisfied with how this works now.

File: llvm/lib/Target/MOS/MOSInstrPseudos.td [hunk 1]
```diff
@@ -311,6 +316,13 @@ def DecMB : MOSIncDecMB {
   dag OutOperandList = (outs GPR:$scratch, variable_ops);
   let Constraints = "@earlyclobber $scratch";
 }
+// DCP-based multi-byte decrement for 6502X. DCP is significantly more compact
```
Comment (mysterymath) [inline llvm/lib/Target/MOS/MOSInstrPseudos.td:319]: It doesn't look like there's any appreciable differences between the surface area of this and the existing DecMB pseudo. What prevents this from being an alternate way of lowering DecMB when 6502X is available?

Comment (asiekierka) [reply llvm/lib/Target/MOS/MOSInstrPseudos.td:319]: For DCP, the scratch register *must* be A: there is no DEC+CPX or DEC+CPY unofficial combination opcode, only DEC+CMP (A).

Comment (mysterymath) [reply llvm/lib/Target/MOS/MOSInstrPseudos.td:319]: Yep, that's what I was missing, thanks.

File: llvm/test/CodeGen/MOS/inc-dec-6502x.ll [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  %0 = add i16 %a, -1
+  ret i16 %0
+}
+
+define i32 @dec_i32(i32 %a) {
+; CHECK-LABEL: dec_i32:
+; CHECK:       ; %bb.0: ; %entry
+; CHECK-NEXT:    tay
+; CHECK-NEXT:    dey
+; CHECK-NEXT:    cmp #0
```
[the remark this answers is not in the document]
Comment (asiekierka) [reply llvm/test/CodeGen/MOS/inc-dec-6502x.ll:122]: When writing the patch, I briefly misinterpeted the code as re-using the existing GPR somehow, and then I forgot to re-review it when I preserved the original behaviour for GPRs. Good point, will fix.

[the remark this answers is not in the document]
Comment (asiekierka) [reply llvm/test/CodeGen/MOS/inc-dec-6502x.ll:122]: (While a transfer is one byte smaller, technically, it clobbers an extra GPR.)

File: llvm/lib/Target/MOS/MOSISelLowering.cpp [hunk 3]
```diff
@@ -409,43 +410,90 @@ static MachineBasicBlock *emitIncDecMB(MachineInstr &MI,
...
+  if (IsDec && !UseDcpOpcode) {
+    // 2/3. DEC: Emit CMP.
+    if (IsReg && !IsMemReg) {
+      Builder.buildInstr(MOS::CMPImm)
+          .addDef(MOS::C)
+          .addUse(MI.getOperand(0).getReg())
+          .addImm(INT64_C(0))
+          .addDef(MOS::Z, RegState::Implicit);
+    } else {
+      Builder.buildInstr(MOS::CMPAbs)
```
[the remark this answers is not in the document]
Comment (asiekierka) [reply llvm/lib/Target/MOS/MOSISelLowering.cpp:496]: Fixed.

File: llvm/test/CodeGen/MOS/inc-dec-6502x.ll (hunk 2 above)
[the remark this answers is not in the document]
Comment (asiekierka) [reply llvm/test/CodeGen/MOS/inc-dec-6502x.ll]: From what I gathered of the original code, the reason was consistency and thus simplicity: the `ld? value / de? value / cp? #0` pattern generalized well to all three cases: GPR, imaginary register, and memory. Now, all three cases are considered separately, in addition to using `#$ff` as the constant as opposed to `#0`.

Either way, I fixed it, but only partially - the scratch register is no longer used in such a case, but it's still allocated. I'm not sure if this is fully solvable well, but the code generation is improved regardless - see diff.

**Review by mysterymath — APPROVED** [review]
Awesome, LGTM!

**Pull request merged**